src/CodeGen_C.cpp

```diff
@@ -181,7 +179,38 @@ const string preamble =
     "}\n";
 }
 
-CodeGen_C::CodeGen_C(ostream &s) : IRPrinter(s), id("$$ BAD ID $$") {}
+CodeGen_C::CodeGen_C(ostream &s, bool is_header, const std::string &guard) : IRPrinter(s), id("$$ BAD ID $$"), is_header(is_header) {
+    if (is_header) {
+        // If it's a header, emit an include guard.
+        stream << "#ifndef HALIDE_" << print_name(guard) << '\n'
+               << "#define HALIDE_" << print_name(guard) << '\n';
+    }
+
+    stream << "#ifdef __cplusplus\n";
+    stream << "#define HALIDE_EXTERN_C extern \"C\"\n";
+    stream << "#else\n";
+    stream << "#define HALIDE_EXTERN_C\n";
```

Comment:
(Let me be elaborate a bit: for some reason people seem to really like the idea of using C macros to define their own programming language to decorate function declarations. I find this makes code much more difficult to read and it makes my life harder because the code might compile one way or the other and I have to think to figure out what is actually happening. It is sometimes necessary to do the macro thing because design of C/C++ is lacking in certain areas, but in this case I feel it can be avoided with no cost. The intent is the functions are always compiled with C linkage and the standard block scoping construct makes this clear.)